parlai/agents/transformer/polyencoder.py

```diff
@@ -0,0 +1,289 @@
+        elif self.attention_type == 'basic':
+            self.attention = BasicAttention(dim=2, attn='basic', get_weights=False)
+
+    def get_encoder(self, opt, dict, null_idx, reduction_type):
+        n_positions = get_n_positions_from_options(opt)
+        embeddings = torch.nn.Embedding(
+            len(dict),
+            opt['embedding_size'],
+            padding_idx=null_idx
+        )
+        torch.nn.init.normal_(embeddings.weight, 0,
+                              opt['embedding_size'] ** -0.5)
+        return TransformerEncoder(
+            n_heads=opt['n_heads'],
+            n_layers=opt['n_layers'],
+            embedding_size=opt['embedding_size'],
+            ffn_size=opt['ffn_size'],
+            vocabulary_size=len(dict),
+            embedding=embeddings,
+            dropout=opt['dropout'],
+            attention_dropout=opt['attention_dropout'],
+            relu_dropout=opt['relu_dropout'],
+            padding_idx=null_idx,
+            learn_positional_embeddings=opt['learn_positional_embeddings'],
+            embeddings_scale=opt['embeddings_scale'],
+            reduction_type=reduction_type,
+            n_positions=n_positions,
+            n_segments=2,
+            activation=opt['activation'],
+            variant=opt['variant'],
+            output_scaling=opt['output_scaling'])
+
+    def encode(self, ctxt_tokens, cand_tokens):
+        """
+            :param ctxt_tokens:
+                2D long tensor, batchsize x sent_len
+            :param cand_tokens:
+                3D long tensor, batchsize x num_cands x sent_len
+                Note this will actually view it as a 2D tensor
+            :returns: (ctxt_rep, ctxt_mask, cand_rep)
+                - ctxt_rep 3D float tensor, batchsize x n_codes x dim
+                - ctxt_mask byte:  batchsize x n_codes (all 1 in case
+                    of polyencoder with code. Which are the vectors to use
+                    in the ctxt_rep)
+                - cand_rep (3D float tensor) batchsize x num_cands x dim
+        """
+        if ctxt_tokens is not None:
+            assert len(ctxt_tokens.shape) == 2
+            bsz = ctxt_tokens.size(0)
+        if cand_tokens is not None:
+            assert len(cand_tokens.shape) == 3
+            bsz = cand_tokens.size(0)
+            num_cands = cand_tokens.size(1)
+
+        cand_embed = None
+        ctxt_rep = None
+        ctxt_rep_mask = None
+
+        if cand_tokens is not None:
+            cand_embed = self.encoder_cand(cand_tokens.view(bsz * num_cands, -1))
+            cand_embed = cand_embed.view(bsz, num_cands, -1)
+
+        if ctxt_tokens is not None:
+            # get context_representation. Now that depends on the cases.
+            ctxt_out, ctxt_mask = self.encoder_ctxt(ctxt_tokens)
+            dim = ctxt_out.size(2)
+
+            if self.type == 'codes':
+                # Basic Attention and MultiHeadAttention share the same API.
```

Comment:
😸

Reply:
Actually got a message from Marie Anne saying that this is not true, I'll modify that